Command.output() no longer respects AbortSignal

Version: Deno 2.7.14

This appears to be a regression introduced in #33335 - by switching to `os_spawn_child` directly, the `signal` property provided via the CommandOptions is ignored.



Minimal test:

```
import { assert, assertFalse } from "jsr:@std/assert";

Deno.test("Deno.Command.output() with signal", async (t) => {
    const signal = AbortSignal.timeout(500);
    const cmd = new Deno.Command("sleep", {
        args: ["10"],
        signal,
    });

    const start = performance.now();

    const output = await cmd.output();

    const duration = performance.now() - start;

    assert(
        duration <= 1000,
        `Command should have been aborted after 500ms, but ran for ${duration.toFixed(2)}ms`,
    );
    assertFalse(output.success);
});
```

Output on 2.7.12 (version predating regression):
```
deno test --allow-run reproduce.test.ts
Check reproduce.test.ts
running 1 test from ./reproduce.test.ts
Deno.Command.output() with signal ... ok (502ms)

ok | 1 passed | 0 failed (505ms)
```

Output on 2.7.14:
```
 eno test --allow-run reproduce.test.ts
Check reproduce.test.ts
running 1 test from ./reproduce.test.ts
Deno.Command.output() with signal ... FAILED (10s)

 ERRORS 

Deno.Command.output() with signal => ./reproduce.test.ts:3:6
error: AssertionError: Command should have been aborted after 500ms, but ran for 10001.12ms
    throw new AssertionError(msg);
          ^
    at assert (https://jsr.io/@std/assert/1.0.19/assert.ts:22:11)
    at file:///home/matt/dev/reelos/reelos/reproduce.test.ts:16:5

 FAILURES 

Deno.Command.output() with signal => ./reproduce.test.ts:3:6

FAILED | 0 passed | 1 failed (10s)

error: Test failed
```

--- a/ext/process/40_process.js
+++ b/ext/process/40_process.js
@@ -617,6 +617,7 @@ function spawnInner(command, {
   env = { __proto__: null },
   uid = undefined,
   gid = undefined,
+  signal = undefined,
   stdin = "null",
   stdout = "piped",
   stderr = "piped",
@@ -655,29 +656,50 @@ function spawnInner(command, {
   const stdoutRid = child.stdoutRid;
   const stderrRid = child.stderrRid;
 
+  let onAbort = null;
+  if (signal !== undefined) {
+    onAbort = () => {
+      try {
+        op_spawn_kill(child.rid, "SIGTERM");
+      } catch {
+        // Ignore the error for https://github.com/denoland/deno/issues/27112
+      }
+    };
+    if (signal.aborted) {
+      onAbort();
+    } else {
+      signal[abortSignal.add](onAbort);
+    }
+  }
+
   return PromisePrototypeThen(
     SafePromiseAll([
       op_spawn_wait(child.rid),
       stdoutRid != null ? readAllRid(stdoutRid) : null,
       stderrRid != null ? readAllRid(stderrRid) : null,
     ]),
-    ({ 0: status, 1: stdout, 2: stderr }) => ({
-      success: status.success,
-      code: status.code,
-      signal: status.signal,
-      get stdout() {
-        if (stdout == null) {
-          throw new TypeError("Cannot get 'stdout': 'stdout' is not piped");
-        }
-        return stdout;
-      },
-      get stderr() {
-        if (stderr == null) {
-          throw new TypeError("Cannot get 'stderr': 'stderr' is not piped");
-        }
-        return stderr;
-      },
-    }),
+    ({ 0: status, 1: stdout, 2: stderr }) => {
+      if (onAbort !== null) {
+        signal[abortSignal.remove](onAbort);
+      }
+      return {
+        success: status.success,
+        code: status.code,
+        signal: status.signal,
+        get stdout() {
+          if (stdout == null) {
+            throw new TypeError("Cannot get 'stdout': 'stdout' is not piped");
+          }
+          return stdout;
+        },
+        get stderr() {
+          if (stderr == null) {
+            throw new TypeError("Cannot get 'stderr': 'stderr' is not piped");
+          }
+          return stderr;
+        },
+      };
+    },
   );
 }
 

--- a/tests/unit/command_test.ts
+++ b/tests/unit/command_test.ts
@@ -471,6 +471,39 @@ Deno.test(
 );
 
 Deno.test(
+  { permissions: { run: true, read: true } },
+  async function commandOutputAbort() {
+    const ac = new AbortController();
+    const command = new Deno.Command(Deno.execPath(), {
+      args: [
+        "eval",
+        "setTimeout(console.log, 1e8)",
+      ],
+      signal: ac.signal,
+      stdout: "null",
+      stderr: "null",
+    });
+    const start = performance.now();
+    queueMicrotask(() => ac.abort());
+    const output = await command.output();
+    const duration = performance.now() - start;
+    assert(
+      duration < 5000,
+      `Expected output() to be aborted promptly, but ran for ${
+        duration.toFixed(2)
+      }ms`,
+    );
+    assertEquals(output.success, false);
+    if (Deno.build.os === "windows") {
+      assertEquals(output.code, 1);
+      assertEquals(output.signal, null);
+    } else {
+      assertEquals(output.code, 143);
+    }
+  },
+);
+
+Deno.test(
   { permissions: { read: true, run: false } },
   async function commandPermissions() {
     await assertRejects(async () => {

